File: python-problems/arrays/src/maximum_subarray_sum.py
def maximum_subarray_sum(nums):
    max_sum = max_ending_here = 0

    for num in nums:
        max_ending_here = max(num, num+max_ending_here)
        max_sum = max(max_ending_here, max_sum)

    return max_sum


def maximum_circular_subarray_sum(nums):
    maximum_subarray_sum_wrap_around = sum(nums) - minimum_subarray_sum(nums)

    return  max(maximum_subarray_sum_wrap_around, maximum_subarray_sum(nums))


def minimum_subarray_sum(nums):
    min_sum = min_ending_here = 0;

    for num in nums:
        min_ending_here = min(num, min_ending_here+num)
        min_sum = min(min_ending_here, min_sum)

    return min_sum

File: java/arrays/src/main/java/MaximumSubarraySum.java
/**
 * Given an array of numbers, find the maximum sum of contiguous subarray of the array
 *
 * eg: Input: [34, -50, 42, 14, -5, 86]
 *
 * Output : 137
 */
public class MaximumSubarraySum {

    public Integer maximumSubarraySumBruteForce(Integer[] array){
        Integer maxSum = 0;

        for(Integer i = 0; i< array.length -1 ; i++){
            Integer sum = 0;
            for(int j=i; j<array.length; j++){
                sum = sum + array[j];
                maxSum = Math.max(sum, maxSum);
            }
        }
        return  maxSum;
    }

    public Integer maximumSubarraySum(Integer[] array){
        Integer maxSum = 0, maxEndingHere = 0;

        for(Integer i=0; i<array.length; i++){
            maxEndingHere = Math.max(array[i], maxEndingHere+array[i]);
            maxSum = Math.max(maxEndingHere, maxSum);
        }
        return maxSum;
    }

    public Integer maximumCircularSubarraySum(Integer[] array){
        Integer maxSum = 0, maxEndingHere = 0, sumOfArray = 0, minSum=0, minEndingHere=0;

        for(Integer i=0; i<array.length;i++){
            sumOfArray+=array[i];
            maxEndingHere = Math.max(array[i], array[i]+maxEndingHere);
            minEndingHere = Math.min(array[i], array[i]+minEndingHere);
            minSum = Math.min(minEndingHere, minSum);
            maxSum = Math.max(maxEndingHere, maxSum);
        }

        return Math.max((sumOfArray-minSum), maxSum);

    }



}
